src/drivers/net/ena.c

```diff
@@ -350,6 +350,89 @@ static int ena_admin ( struct ena_nic *ena, union ena_aq_req *req,
 	return rc;
 }
 
+/**
+ * Set async event notification queue config
+ *
+ * @v ena		ENA device
+ * @ret rc		Return status code
+ */
+static int ena_set_aenq_config ( struct ena_nic *ena,
+				 uint32_t enabled_groups ) {
+	union ena_aq_req *req;
+	union ena_acq_rsp *rsp;
+	union ena_feature *feature;
+	int rc = 0;
+
+	/* Construct request */
+	req = ena_admin_req ( ena );
+	req->header.opcode = ENA_SET_FEATURE;
+	req->set_feature.id = ENA_AENQ_CONFIG;
+	feature = &req->set_feature.feature;
+	feature->aenq.enabled_groups = cpu_to_le32 ( enabled_groups );
+
+	/* Issue request */
+	if ( ( rc = ena_admin ( ena, req, &rsp ) ) != 0 )
+		DBGC ( ena, "ENA %p failed to set AENQ config\n", ena );
+
+	return rc;
+}
+
+
+/**
+ * Create async queues
+ *
+ * @v ena		ENA device
+ * @ret rc		Return status code
+ */
+static int ena_create_async ( struct ena_nic *ena ) {
+	size_t aenq_len = ( ENA_AENQ_COUNT * sizeof ( ena->aenq.evt[0] ) );
+	int rc;
+
+	/* Allocate async event queue */
+	ena->aenq.evt = malloc_phys ( aenq_len, aenq_len );
+	if ( ! ena->aenq.evt ) {
+		rc = -ENOMEM;
+		goto err_alloc_aenq;
+	}
+	memset ( ena->aenq.evt, 0, aenq_len );
+
+	/* Program queue addresses and capabilities */
+	ena_set_base ( ena, ENA_AENQ_BASE, ena->aenq.evt );
+	ena_set_caps ( ena, ENA_AENQ_CAPS, ENA_AENQ_COUNT,
+		       sizeof ( ena->aenq.evt[0] ) );
+
+	DBGC ( ena, "ENA %p AENQ [%08lx,%08lx)\n",
+	       ena, virt_to_phys ( ena->aenq.evt ),
+	       ( virt_to_phys ( ena->aenq.evt ) + aenq_len ) );
+
+	/* Disable all events. That way we don't need to handle the queue. */
+	rc = ena_set_aenq_config ( ena, 0 );
+	if ( !rc )
+		return 0;
```

Comment:
Please follow the general pattern:

```
if ( ( rc = try_thing() ) != 0 )
   goto err_thing;
```

followed by a standalone

```
return 0;
```

when the flow of execution successfully reaches the end of the function, followed by

```
    undo_thing();
err_thing:
```

(In this case, I'm guessing that there won't be any `undo_thing()`, since there is nothing to undo from an attempt to disable events.)